shared/api/url.go

```diff
@@ -0,0 +1,30 @@
+package api
+
+import (
+	"net/url"
+	"strings"
+)
+
+// ResourceURL helps to safely generate a URL to a LXD resource.
+// It starts with a "/" and then appends each of the pathParts (escaped using url.PathEscape) separated with a "/".
+// The projectName will be added as an escaped query parameter "project" if it is non-empty or not "default".
+func ResourceURL(projectName string, pathParts ...string) *url.URL {
```

Comment:
@stgraber another approach that may be more user friendly is to define a new interface called `api.Addresser` that has requires a `URL() string` function. Then update this function's signature to be `ResourceURL(projectName string, addresser Addresser) *url.URL`

We would then add a `URL() string` function to each resource type in `api` package, that would return its resource URL, e.g. `/1.0/networks/lxdbr0`.

We would probably need another function such as `URLBuild()` that has the same signature as `ResourceURL` does now, which each resource type could use to implements its `URL()` function.

One concern I did have though is that would require importing `shared/version` into `shared/api`, would that be acceptable?

Reply:
I've renamed `ResourceURL` to `URLBuild` in case we do add `ResourceURL` as well.

Reply:
Shouldn't it rather be named `BuildURL` as it actually performs a task?

Reply:
I tend to use more and more the topic first then action, as it helps with auto complete in editor and often we end up adding extra related functionality and prefixing with topic first keeps them together. I can see my proposed Resource URL becoming URLOfResource() for example.